crates/schema/src/type_for_generate.rs

```diff
@@ -0,0 +1,516 @@
+//! `AlgebraicType` extensions for generating client code.
+
+use spacetimedb_data_structures::{
+    error_stream::{CollectAllErrors, CombineErrors, ErrorStream},
+    map::{HashMap, HashSet},
+};
+use spacetimedb_lib::{AlgebraicType, ProductTypeElement};
+use spacetimedb_sats::{typespace::TypeRefError, AlgebraicTypeRef, ArrayType, SumTypeVariant, Typespace};
+use std::sync::Arc;
+
+use crate::{
+    error::{IdentifierError, PrettyAlgebraicType},
+    identifier::Identifier,
+};
+
+/// Errors that can occur when rearranging types for client codegen.
+#[derive(thiserror::Error, Debug, PartialOrd, Ord, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ClientCodegenError {
+    #[error(
+        "internal codegen error: non-special product or sum type {ty} cannot be used to generate a client type use"
+    )]
+    NonSpecialTypeNotAUse { ty: PrettyAlgebraicType },
+
+    #[error("internal codegen error: invalid AlgebraicTypeRef")]
+    TypeRefError(#[from] TypeRefError),
+
+    #[error("internal codegen error: type ref {ref_} was not pre-declared as a definition")]
+    NonDeclaredTypeDef { ref_: AlgebraicTypeRef },
+
+    #[error("internal codegen error: all type elements require names: {ty}")]
```

Comment:
I'd reword this, "type elements" sounds like talking about elements of the type (ProductTypeElement, SumTypeVariant), not types themselves, and can be confusing.

Reply:
Ah, in fact that was the original error here but the error message persisted through a refactor.

Reply:
Wait, the enum variant is confusingly named, the error message was right. Renamed.